### PR TITLE
Transition to using `access_control_list` Vespa field for permissioning

### PR DESCRIPTION
**Pull request by Weves**
Lays the framework for adding document sets.

- Removes `allowed_users` / `allowed_groups` from document indexes and uses `access_control_list` instead
- Simplifies document deletion + makes all document updates safe for multi-processing
- Adds `access` utility functions, this "module" be where all the logic related to computing a users access levels lives

### CODE REVIEW

File: backend/danswer/access/access.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            document_sets=set(
+                [build_personal_document_set_name(user_id) for user_id in user_ids]
+                + ([PUBLIC_DOCUMENT_SET] if is_public else [])
+            ),
+        )
+        for document_id, user_ids, is_public in document_access_info
+    }
+
+
+def get_access_for_documents(
```
Comment (yuhongsun96) [inline backend/danswer/access/access.py:25]: :( why do you like these extra abstractions, ew

File: backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd [hunk 2]
```diff
@@ -64,6 +64,10 @@ schema danswer_chunk {
             indexing: summary | attribute
             attribute: fast-search
         }
+        field document_sets type weightedset<string> {
```
Comment (yuhongsun96) [inline backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd:63]: I honestly would just prefer deleting the users/groups, if it breaks so be it... let's not keep lots of junk around

File: backend/danswer/db/document.py [hunk 3]
```diff
@@ -10,73 +12,64 @@
...
     db_session: Session,
-    connector_id: int,
-    credential_id: int,
-) -> Sequence[DocumentByConnectorCredentialPair]:
-    initial_doc_ids_stmt = select(DocumentByConnectorCredentialPair.id).where(
-        and_(
-            DocumentByConnectorCredentialPair.connector_id == connector_id,
-            DocumentByConnectorCredentialPair.credential_id == credential_id,
+    document_ids: list[str],
+) -> Sequence[tuple[str, int]]:
```
Comment (yuhongsun96) [inline backend/danswer/db/document.py:43]: some comment/description please

File: backend/danswer/server/document_sets.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+from fastapi import Depends
+
+from danswer.auth.users import current_admin_user
+from danswer.db.models import User
+
+
+router = APIRouter(prefix="/manage")
+
+
+@router.patch("/promote-user-to-admin")
```
Comment (yuhongsun96) [inline backend/danswer/server/document_sets.py]: nice meme

File: backend/danswer/server/document_sets.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from fastapi import APIRouter
+from fastapi import Depends
+
+from danswer.auth.users import current_admin_user
+from danswer.db.models import User
+
+
+router = APIRouter(prefix="/manage")
```
Comment (yuhongsun96) [inline backend/danswer/server/document_sets.py]: is this still best?

File: backend/danswer/db/document.py [hunk 6]
```diff
@@ -187,3 +178,40 @@ def delete_documents_complete(db_session: Session, document_ids: list[str]) -> N
     )
     delete_documents(db_session, document_ids)
     db_session.commit()
+
+
+def acquire_document_locks(db_session: Session, document_ids: list[str]) -> bool:
```
Comment (yuhongsun96) [inline backend/danswer/db/document.py:201]: Can we make a comment on the batch size and the worse case conflict flows?

File: backend/danswer/db/document.py [hunk 7]
```diff
@@ -10,73 +12,64 @@
...
 
-    # Filter it down to the documents with more than 1 connector/credential pair
-    # Meaning if this connector/credential pair is removed, this doc is still accessible
-    trimmed_doc_ids_stmt = (
-        select(DbDocument.id)
+
+def get_acccess_info_for_documents(
+    db_session: Session,
+    document_ids: list[str],
+) -> Sequence[tuple[str, list[UUID | None], bool]]:
```
Comment (yuhongsun96) [inline backend/danswer/db/document.py:59]: gets the DocumentByConnectorCredentialPair and the users who have access to it? and if its public? can we add some comments to this stuff?

File: backend/danswer/datastores/vespa/store.py [hunk 8]
```diff
@@ -246,16 +213,17 @@ def _index_chunk(
...
-    if user_id:
-        filter_str += (
-            f'({ALLOWED_USERS} contains "{user_id}" or '
-            f'{ALLOWED_USERS} contains "{PUBLIC_DOC_PAT}") and '
-        )
-    else:
-        filter_str += f'{ALLOWED_USERS} contains "{PUBLIC_DOC_PAT}" and '
+    # TODO: have document sets passed in directly
+    document_sets = get_document_sets_for_user(user_id=user_id)
+    document_set_filter_stmts = [
```
Comment (yuhongsun96) [inline backend/danswer/datastores/vespa/store.py]: Any performance considerations here? It's one of those many compare many situations, potentially quite slow no?

File: backend/danswer/access/access.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        user_ids=access.user_ids
+        if user_id is None
+        else access.user_ids.union({user_id}),
+        document_sets=access.document_sets
+        if document_sets is None
+        else document_sets.union(document_sets),
+    )
+
+
+def access_identical(access1: Access, access2: Access) -> bool:
```
Comment (yuhongsun96) [inline backend/danswer/access/access.py]: seems unused

File: backend/danswer/access/document_set_utils.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+from uuid import UUID
+
+from danswer.configs.constants import PUBLIC_DOCUMENT_SET
+
+
+def build_personal_document_set_name(user_id: UUID | None) -> str:
+    if user_id is None:
+        return PUBLIC_DOCUMENT_SET
+
+    return f"__USER_{user_id}"
```
Comment (yuhongsun96) [inline backend/danswer/access/document_set_utils.py]: nit: prefer having this be PERSONAL_DOCUMENT_SET.format(user_id) or something for consistency

File: backend/danswer/access/access.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    document_ids: list[str], db_session: Session | None = None
+) -> dict[str, Access]:
+    if db_session is None:
+        with Session(get_sqlalchemy_engine()) as db_session:
+            return _get_access_for_documents(document_ids, db_session)
+
+    return _get_access_for_documents(document_ids, db_session)
+
+
+def update_access(
```
Comment (yuhongsun96) [inline backend/danswer/access/access.py]: seems unused

File: backend/danswer/background/connector_deletion.py [hunk 12]
```diff
@@ -10,184 +10,130 @@
...
 from danswer.db.engine import get_sqlalchemy_engine
 from danswer.db.index_attempt import delete_index_attempts
-from danswer.db.models import ConnectorCredentialPair
-from danswer.db.models import Credential
+from danswer.server.models import ConnectorCredentialPairIdentifier
 from danswer.utils.logger import setup_logger
 
 logger = setup_logger()
 
+_DELETION_BATCH_SIZE = 1000
```
Comment (yuhongsun96) [inline backend/danswer/background/connector_deletion.py:35]: If deletion is much much faster and has larger batches, should we include logic for deletion to priority grab/hold locks until it has all of the necessary ones to run?

Maybe this is hard to implement

File: backend/danswer/datastores/vespa/store.py [hunk 13]
```diff
@@ -183,12 +152,10 @@ def _index_vespa_chunks(
             METADATA: json.dumps(document.metadata),
             EMBEDDINGS: embeddings_name_vector_map,
             BOOST: DEFAULT_BOOST,
-            ALLOWED_USERS: cross_connector_document_metadata_map[
-                document.id
-            ].allowed_users,
-            ALLOWED_GROUPS: cross_connector_document_metadata_map[
-                document.id
-            ].allowed_user_groups,
+            # give each document set "equal weighting" in the weightedset
```
Comment (yuhongsun96) [inline backend/danswer/datastores/vespa/store.py]: Is this for upcoming feature of document-sets boosting?

If this is supported, how will we handle different personas or knowledge sets favoring different document sets?

Comment (yuhongsun96) [reply backend/danswer/datastores/vespa/store.py]: for future clarity, let's call them personas which have document sets (knowledge sets as a term is banned!)

**Review by yuhongsun96 — CHANGES_REQUESTED** [review]

File: backend/danswer/datastores/typesense/store.py [hunk 14]
```diff
@@ -181,12 +139,7 @@ def _index_typesense_chunks(
                 SOURCE_LINKS: json.dumps(chunk.source_links),
                 SEMANTIC_IDENTIFIER: document.semantic_identifier,
                 SECTION_CONTINUATION: chunk.section_continuation,
-                ALLOWED_USERS: cross_connector_document_metadata_map[
-                    document.id
-                ].allowed_users,
-                ALLOWED_GROUPS: cross_connector_document_metadata_map[
-                    document.id
-                ].allowed_user_groups,
+                DOCUMENT_SETS: json.dumps(list(chunk.access.document_sets)),
```
Comment (Weves) [inline backend/danswer/datastores/typesense/store.py]: add comment about performance issue

File: backend/danswer/chunking/models.py [hunk 15]
```diff
@@ -55,6 +58,34 @@ class IndexChunk(DocAwareChunk):
     embeddings: ChunkEmbedding
 
 
+@dataclass
+class DocMetadataAwareIndexChunk(IndexChunk):
```
Comment (Weves) [inline backend/danswer/chunking/models.py:62]: Note: could also attach this info to the document itself, but this doesn't play as well with our current setup

File: backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd (hunk 2 above)
Comment (Weves) [reply backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd:63]: removed!

File: backend/danswer/chunking/models.py (hunk 15 above)
Comment (yuhongsun96) [reply backend/danswer/chunking/models.py:62]: Instead of the chunk or?

File: backend/scripts/migrate_vespa_to_acl.py [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+logger = setup_logger()
+
+
+def _migrate_vespa_to_acl() -> None:
+    vespa_index = get_default_document_index()
+    if not isinstance(vespa_index, VespaIndex):
+        raise ValueError("This script is only for Vespa indexes")
+
+    with Session(get_sqlalchemy_engine()) as db_session:
+        # for all documents, set the `access_control_list` field apporpriately
```
Comment (yuhongsun96) [inline backend/scripts/migrate_vespa_to_acl.py:25]: typo lul

File: backend/scripts/migrate_vespa_to_acl.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+        # based on the state of Postgres
+        documents = db_session.scalars(select(Document)).all()
+        document_access_info = get_acccess_info_for_documents(
+            db_session=db_session,
+            document_ids=[document.id for document in documents],
+        )
+        vespa_index.update(
```
Comment (yuhongsun96) [inline backend/scripts/migrate_vespa_to_acl.py:32]: I would prefer if there's a way to also erase the no longer relevant fields but... if there's no way without breaking then whatever

Comment (yuhongsun96) [reply backend/scripts/migrate_vespa_to_acl.py:32]: https://docs.vespa.ai/en/schemas.html#schema-modifications

Nobody is using private docs atm anyway so we can probably just remove/rename things and just stick public into every doc?

Comment (yuhongsun96) [reply backend/scripts/migrate_vespa_to_acl.py:32]: But also, people who dont have the highlighting feature should reindex anyway

Comment (Weves) [reply backend/scripts/migrate_vespa_to_acl.py:32]: no

**Comment by yuhongsun96** [comment]
Is this still relevant? or is this covered by the other PR now?

**Pull request merged**